### ISSUE TITLE
Bug: in react-devtools-inline case <ReactDevtools/> closing bridge too aggressively

### ISSUE THREAD

**Issue by sean9keenan** [facebook/react#20135]
DevTools closes the bridge whenever it is unmounted [(Link to relevant code)](https://github.com/facebook/react/blob/master/packages/react-devtools-shared/src/devtools/views/DevTools.js#L172-L181). In the case of the chrome extension this is fine because the node is never unmounted, but in use cases with `react-devtools-inline` (eg: codesandbox), where you are only given one <DevTools/> instance this causes the bridge to be closed, and never recreated if for whatever reason <DevTools/> is unmounted (which happens when switching tabs).

I am willing to submit a PR, but wanted to discuss the fix before making it - the fix I see is adding a flag to <DevTools/> to not close the bridge on unmount, and making it an opt-in flag users of <DevTools/> (recommended in the case of react-devtools-inline). Adding an ability to manually access the bridge would then let the developer manually control the lifecycle - but I'm unsure the risk/issue of keeping a dangling bridge open, so I don't see this as strictly necessary.

React version: N/A
React-devtools-inline: 4.9.0

## Steps To Reproduce

1.  Create a code-sandbox React environment
2. Open React DevTools tab
3. Switch to Problems tab (or any other tab)
4. Switch back to React DevTools tab
5. Bridge has been closed, `index.js:55 Cannot send message "highlightNativeElement" through a Bridge that has been shutdown.` is printed in console. New react nodes can not be inspected.

Link to code example: N/A
Code example isn't relevant, any CodeSandbox will reproduce this. (even with no additional code)

## The current behavior
Inspecting elements that haven't previously been inspected will fail because bridge communication fails if the <DevTools/> ever unmounts and remounts

## The expected behavior
The bridge doesn't close even on unmount when using `react-devtools-inline`, or an option is available when mounting <DevTools/> to not automatically close the bridge.

**Comment by bvaughn** [comment]
I'm not super comfortable with the idea of leaving the bridge open (or making an option to leave it open). I wonder if we couldn't either:
* Hide (rather than unmount) the DevTools if its tab isn't active, or...
* Just throw-away and recreate DevTools (including the Bridge) when it gets unmounted and shown again.

**Comment by sean9keenan** [comment]
Makes sense! I wasn't aware there was much of a risk to leaving the bridge open.

Option 1) is straightforward - if this is recommended, do you mind if I add some documentation for it?

**Comment by bvaughn** [comment]
> Is it possible to reopen a closed bridge?

Not really. Shutting down the Bridge sends a signal through to the other side telling it to also shutdown. That means there's no longer anything left "listening" to know about a "reopen". (That's the point, to remove listeners so we don't leak.)

> Option 1) is straightforward - if this is recommended, do you mind if I add some documentation for it?

Sure. 😄 

> Option 2) also seems like it could make sense - although I'm not 100% sure how to best do this in the context of react-devtools-inline

The inline package gets used like this:
```js
const DevTools = initializeFrontend(contentWindow);
ReactDOM.render(<DevTools {..otherProps} />, container);
```

If you later unmount the UI, you should just throw it away so that if it's re-mounted again it's also recreated (most importantly– the Bridge is recreated).

Looks like Code Sandbox [already uses a Wrapper component](https://github.com/codesandbox/codesandbox-client/blob/dc1d54d5b804d3deee35f39fe503cc1833fe0c01/packages/app/src/app/components/Preview/DevTools/React-Devtools/index.tsx) to create and mount DevTools, so that thing should just cleanup on unmount?

Maybe something like this:
```js
function ExampleWrapper() {
  const devToolsRef = useRef(null);
  let ReactDevTools = devToolsRef.current;
  if (ReactDevTools === null) {
    // Create/recreate DevTools (with a new Bridge).
    devToolsRef.current = ReactDevTools = initialize(contentWindow);
  }

  useEffect(() => {
    return () => {
      // Destroy DevTools (and the disconnected Bridge) on unmount.
      devToolsRef.current = null;
    };
  }, []);

  return (
    <ReactDevTools
      viewElementSourceFunction={viewElementSourceFunction}
      browserTheme={browserTheme}
    />
  );
}
```

Or just lift the DevTools initialization call out of the Wrapper component entirely and pass the configured component down as a prop (which is what I'd probably have done).

**Comment by sean9keenan** [comment]
> > Is it possible to reopen a closed bridge?
> 
> Not really. Shutting down the Bridge sends a signal through to the other side telling it to also shutdown. That means there's no longer anything left "listening" to know about a "reopen". (That's the point, to remove listeners so we don't leak.)

Good to know, thanks for clarifying!

> > Option 1) is straightforward - if this is recommended, do you mind if I add some documentation for it?
> 
> Sure. 😄

Awesome! I'll plan on trying to get to that this weekend.

> > Option 2) also seems like it could make sense - although I'm not 100% sure how to best do this in the context of react-devtools-inline
>
>  (... removed some of above quote)
> Or just lift the DevTools initialization call out of the Wrapper component entirely and pass the configured component down as a prop (which is what I'd probably have done).

Makes sense - I'll see how easy it will be to get a PR into codesandbox, that actually sounds like a pretty neat task to get involved over there! Not actually a maintainer or anything, just noticed they were running into the same problem I was - so an easy thing to point out as a repro 😄.

Thanks again for your help!

**Comment by Ankit834** [comment]
@bvaughn Can you help me with similar kind of issue:-
We are using Sandboxed iframe, there we have multiple tabs under devtools and on Tab Switch, the ReactDevtools gets unmounted where we are actually activating the backend.
Now when we initialize back the ReactDevtool on component mount, it is not working anymore.
Steps:-
- In iframe.html => doing same as suggested here 
```
import { activate, initialize } from "react-devtools-inline/backend";

// The DevTools hook needs to be installed before React is even required!
// The safest way to do this is probably to install it in a separate script tag.
initialize(window);

// Wait for the frontend to let us know that it's ready.
function onMessage({ data }) {
  switch (data.type) {
    case "activate-backend":
      window.removeEventListener("message", onMessage);

      activate(window);
      break;
    default:
      break;
  }
}

window.addEventListener("message", onMessage);
```  
- the in main-window.html on iframe load, posting message to 'activate-backend' and rendering the component. Now when switching the tab, the component unmounts and so the bridge is closed as per my understanding.
When I try to initialize the Devtool again, it is not working.
So how can it be achieved?
And is it correct way to activate backend again everytime when I switch back to ReactDevtool component along with Devtool initialization?
